sdk/nodejs/output.ts

```diff
@@ -184,7 +187,9 @@ This function may throw in a future version of @pulumi/pulumi.`;
             // The known state of the output we're returning depends on if we're known as well, and
             // if a potential lifted inner Output is known.  If we get an inner Output, and it is
             // not known itself, then the result we return should not be known.
-            const resultIsKnown = Promise.all([isKnown, innerDetails]).then(([k1, k2]) => k1 && k2.isKnown);
+            const resultIsKnown = Promise.all([isKnown, innerDetails]).then(([k1, k2]) => {
+                return (k1 || runWithUnknowns) && k2.isKnown;
```

Comment:
why this and not: `(k1 && k2.isKnown) || runWithUnknowns`?   And, regardless of hte reason, please document why this is the desired semantics here.

Reply:
> why this and not: (k1 && k2.isKnown) || runWithUnknowns?

These two expressions are not equivalent. The expression in the code reads "the result is known if the inner `isKnown` resolves to `true` and either the outer `isKnown` resolves to `true` or if we are explicitly processing unknown values. Your suggestion reads "the result is known if the inner `isKnown` and the outer `isKnown` both resolve to `true` or if we are explicitly processing unknown values. The former only allows the result to be known if the inner result is known; the latter allows the result to be known even if the inner result is unknown.

Reply:
> These two expressions are not equivalent. 

i agree they're not equivalent.  :)   my point was that it wasn't clear why one result was the desirable one over the other.

Regardless, my overall point is the necessity to document in the code as the code itself is not readily apparent reading it directly.  If `(k1 || runWithUnknowns) && k2.isKnown` are the semantics desired, we should document clearly why that is.  

thanks :)

Reply:
Note: in cases like this, i think it's super valuable to have examples *in the comment* demonstrating why this produces a behavior desired versus alternative approaches.  It gives later readers a good demonstration of why something is desirable versus leaving them both having to ask "is this intentional or a bug?" as well as having to try to explicitly enumerate the appropriate set of cases mentally to convince themselves.

Reply:
Improved the leading comment.